lxd/instance/drivers/driver_qemu.go

```diff
@@ -1018,7 +1018,7 @@ func (d *qemu) Start(stateful bool) error {
 	}
 
 	// Ensure secureboot is turned off for images that are not secureboot enabled
-	if shared.IsFalse(d.localConfig["image.requirements.secureboot"]) && !shared.IsFalse(d.expandedConfig["security.secureboot"]) {
+	if !shared.IsTrueOrEmpty(d.localConfig["image.requirements.secureboot"]) && shared.IsTrueOrEmpty(d.expandedConfig["security.secureboot"]) {
```

Comment:
The first part of this check wants to specifically know if `image.requirements.secureboot` is set to `false`, `no` or `0`.
If it's set to `foo`, it should not return true.

Reply:
My understanding was that it was testing if `image.requirements.secureboot` was *logically* false, which AFAIK everywhere else in the LXD code based is equivalent to `!IsTrue()` (with additional logic to cater for whether the setting is true if empty on a case by case basis).

So anywhere that we use `shared.IsTrue()` means that "foo" is considered logically false.
Except here where "foo" is considered logically true (i.e not `!IsFalse()`).

This is the inconsistency I was trying to remove. 
Of course I'm also assuming that the validation is preventing it being set to "foo" anyway.

Reply:
If `image.requirements.secureboot` can be set to "foo" and there's no validation yet, then perhaps we also need to improve that so that the image isn't even allowed to be used if it has invalid meta data.

Reply:
Probably need to use `validate.IsBool(value)` around here somewhere.

Reply:
There are 64 uses of `!shared.IsTrue()` in the codebase currently. Which is what gave me the impression that it was a common approach to consider the opposite of `IsTrue` to be logical false. 

I've also seen a fair few examples of `shared.IsTrue(d.config["some.key"]) || d.config["some.key"] == ""` which the new function `IsTrueOrEmpty` was make to make easier and shorter to read.

I am concerned about having `IsFalse*()` functions at all because they are going to invite these sorts of logic "gaps" (where some parts of the system consider a value true and others consider the same value false).

But if I'm understanding you correctly, the `image.requirements.secureboot` setting can be missing/empty, and for backwards compatibility we should only perform the check if its been explicitly supplied.

In which case can we not just add an additional condition to that section that the `image.requirements.secureboot` is non-empty or even passing the `validate.IsBool()` check, before then using `shared.IsTrue`?

This feels like it would make it more explicit that we expect the value to be set before considering it.

Reply:
The `!shared.IsTrue` in many cases came from us not having `shared.IsFalse` and being lazy about it ;)

I'd actually like to see all `!shared.Is*` abolished in favor of explicit `shared.IsTrue`, `shared.IsFalse` or their `OrEmpty` variants. This will prevent surprises should we ever change one of those boolean fields into some kind of tri-state or similar value set.

Reply:
Fair enough. In that case I think we need those functions to return an error when they receive an unexpected value, to avoid the same unexpected value being treated as two different logical values depending on the function call made by the developer.

Perhaps we should have a unified `ToBool(value string, empty bool) (bool, bool)` function that can be used as follows:

```go
if enabled, valid := ToBool("foo", true); valid && enabled {
    // We are confident of the config value we have allowed to get here.
}
``

Reply:
Based on our catch up discussion the plan is to have IsTrue, IsTrueOrEmpy, IsFalse and IsFalseOrEmpty, and then importantly, remove all uses of `!IsTrue()` and replace with `IsFalse()`.